src/Microsoft.AspNetCore.Authentication2/IAuthenticationSchemeHandler.cs

```diff
@@ -0,0 +1,212 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Text.Encodings.Web;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Internal;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Authentication2
+{
+    // Created on a per request basis to handle one particular scheme.
+    public interface IAuthenticationSchemeHandler
+    {
+        // Gives the handler access to the configuration data
+        Task InitializeAsync(AuthenticationScheme scheme, HttpContext context);
+
+        Task AuthenticateAsync(AuthenticateContext context);
+        Task ChallengeAsync(ChallengeContext context);
+        Task SignInAsync(SignInContext context);
+        Task SignOutAsync(SignOutContext context);
+    }
+
+    public abstract class AuthenticationSchemeHandler<TOptions> : IAuthenticationSchemeHandler where TOptions : class
```

Comment:
Perf: How can we avoid re-creating these handlers on every request? Is there some way we can make them stateless and store the relevant state in the HttpContext?